controllers/volumereplicationgroup_controller.go

```diff
@@ -1145,6 +1173,103 @@ func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName, state volr
 	return nil
 }
 
+// namespacedName applies to both VolumeReplication resource and pvc as of now.
+// This is because, VolumeReplication resource for a pvc that is created by the
+// VolumeReplicationGroup has the same name as pvc. But in future if it changes
+// functions to be changed would be processVRAsPrimary(), processVRAsSecondary()
+// to either receive pvc NamespacedName or pvc itself as an additional argument.
+func (v *VRGInstance) selectVolumeReplicationClass(namespacedName types.NamespacedName) (string, error) {
+	className := ""
+
+	if !v.vrcUpdated {
+		if err := v.updateReplicationClassList(); err != nil {
+			v.log.Error(err, "Failed to get VolumeReplicationClass list")
+
+			return className, fmt.Errorf("failed to get VolumeReplicationClass list")
+		}
+
+		v.vrcUpdated = true
+	}
+
+	if len(v.replClassList.Items) == 0 {
+		v.log.Info("No VolumeReplicationClass available")
+
+		return className, fmt.Errorf("no VolumeReplicationClass available")
+	}
+
+	storageClass, err := v.getStorageClass(namespacedName)
+	if err != nil {
+		v.log.Info(fmt.Sprintf("Failed to get the storageclass of pvc %s",
+			namespacedName))
+
+		return className, fmt.Errorf("failed to get the storageclass of pvc %s (%w)",
+			namespacedName, err)
+	}
+
+	for index := range v.replClassList.Items {
+		replicationClass := &v.replClassList.Items[index]
+		if storageClass.Provisioner != replicationClass.Spec.Provisioner {
+			continue
+		}
+
+		schedule, found := replicationClass.Spec.Parameters["schedule"]
+		if !found {
+			// schedule not present in parameters of this replicationClass.
+			continue
+		}
+
+		// ReplicationClass that matches both VRG schedule and pvc provisioner
+		if schedule == *v.instance.Spec.Schedule {
+			className = replicationClass.Name
+
+			break
+		}
+	}
+
+	if className == "" {
+		v.log.Info(fmt.Sprintf("No VolumeReplicationClass found to match provisioner and schedule %s/%s",
+			storageClass.Provisioner, *v.instance.Spec.Schedule))
+
+		return className, fmt.Errorf("no VolumeReplicationClass found to match provisioner and schedule")
+	}
+
+	return className, nil
+}
+
+func (v *VRGInstance) getStorageClass(namespacedName types.NamespacedName) (*storagev1.StorageClass, error) {
+	pvc := &corev1.PersistentVolumeClaim{}
+
+	for index := range v.pvcList.Items {
+		pvcItem := &v.pvcList.Items[index]
+
+		pvcNamespacedName := types.NamespacedName{Name: pvcItem.Name, Namespace: pvcItem.Namespace}
+		if pvcNamespacedName == namespacedName {
+			pvc = pvcItem
+
+			break
+		}
+	}
+
+	if pvc == nil {
+		v.log.Info("failed to get the pvc with namespaced name", namespacedName)
+
+		// Need the storage driver of pvc. If pvc is not found return error.
+		return nil, fmt.Errorf("failed to get the pvc with namespaced name %s", namespacedName)
+	}
+
+	scName := pvc.Spec.StorageClassName
+
+	storageClass := &storagev1.StorageClass{}
+	if err := v.reconciler.Get(v.ctx, types.NamespacedName{Name: *scName}, storageClass); err != nil {
```

Comment:
(nit) minor optimization possible here, as most often all PVCs would point to the same SC, so if we stash away the SCs we fetched, we can avoid fetching it again for the next PVC saving a call to the API server. (for later, just noting it for now)

Reply:
ACK.